### PR TITLE
Enable binary shortlist loading from bytebuffer

### PR DESCRIPTION
**Pull request by qianqianzhu**
### Description
This PR enables binary shortlist loading directly from a bytebuffer, which is used to create `shortlist_memory` for `bergamot-translator`.
It is related to issues: https://github.com/browsermt/bergamot-translator/issues/58#issue-836192230

Lists of changes:
- Add a new Binary shortlist constructor: loading directly from a bytebuffer. 

Added dependencies: none

### How to test
Manually tested.

### Checklist

- [x] I have tested the code manually
- [ ] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [ ] I have updated CHANGELOG.md

### CODE REVIEW

File: src/data/shortlist.h [hunk 1]
```diff
@@ -210,6 +210,16 @@ class BinaryShortlistGenerator : public ShortlistGenerator {
...
 
+  // construct directly from buffer
+  BinaryShortlistGenerator(const void* ptr_void,
+                           const size_t blobSize,
+                           Ptr<const Vocab> srcVocab,
+                           Ptr<const Vocab> trgVocab,
+                           size_t srcIdx = 0,
+                           size_t /*trgIdx*/ = 1,
+                           bool shared = false,
+                           bool check = true);
```
Comment (jerinphilip) [inline src/data/shortlist.h:221]: I saw your check is just `srcVocab == trgVocab`, why take this additionally if you have access to both arguments?

Comment (qianqianzhu) [reply src/data/shortlist.h:221]: Yes, this is a new variable that tells whether the binary file needs to be checked or not.

Comment (jerinphilip) [reply src/data/shortlist.h:221]: The query remains, but about `shared`, why do you need the extra argument `shared`, if you have access to both `srcVocab `and `trgVocab`, from the arguments to this function and you can infer it within the function?

Comment (qianqianzhu) [reply src/data/shortlist.h:221]: That is a legacy issue. I kept `shared` there for the concern of backward compatibility.

Comment (jerinphilip) [reply src/data/shortlist.h:221]: Could you leave a note here about the same in comments. Thanks.

Comment (qianqianzhu) [reply src/data/shortlist.h:221]: Sure!

**Comment by jerinphilip** [comment]
@qianqianzhu @kpu can we get this merged?

**Review by kpu — APPROVED** [review]

**Pull request merged**